domain/schema_validator.go

```diff
@@ -137,44 +139,58 @@ func (s *schemaValidator) Update(leaseGrantTS uint64, oldVer, currVer int64, cha
 	// Update the schema deltaItem information.
 	if currVer != oldVer {
 		logutil.BgLogger().Debug("update schema validator", zap.Int64("oldVer", oldVer),
-			zap.Int64("currVer", currVer), zap.Int64s("changedTableIDs", changedTableIDs))
-		s.enqueue(currVer, changedTableIDs)
+			zap.Int64("currVer", currVer), zap.Int64s("changedTableIDs", changedTableIDs),
+			zap.Uint64s("changedActionTypes", changedActionTypes))
+		s.enqueue(currVer, changedTableIDs, changedActionTypes)
 	}
 }
 
-func hasRelatedTableID(relatedTableIDs, updateTableIDs []int64) bool {
-	for _, tblID := range updateTableIDs {
-		for _, relatedTblID := range relatedTableIDs {
-			if tblID == relatedTblID {
-				return true
-			}
-		}
-	}
-	return false
-}
-
 // isRelatedTablesChanged returns the result whether relatedTableIDs is changed
 // from usedVer to the latest schema version.
 // NOTE, this function should be called under lock!
-func (s *schemaValidator) isRelatedTablesChanged(currVer int64, tableIDs []int64) bool {
+func (s *schemaValidator) isRelatedTablesChanged(currVer int64, tableIDs []int64, getAllInfo bool) ([]int64, []uint64, bool) {
 	if len(s.deltaSchemaInfos) == 0 {
 		metrics.LoadSchemaCounter.WithLabelValues(metrics.SchemaValidatorCacheEmpty).Inc()
 		logutil.BgLogger().Info("schema change history is empty", zap.Int64("currVer", currVer))
-		return true
+		return nil, nil, true
 	}
 	newerDeltas := s.findNewerDeltas(currVer)
 	if len(newerDeltas) == len(s.deltaSchemaInfos) {
 		metrics.LoadSchemaCounter.WithLabelValues(metrics.SchemaValidatorCacheMiss).Inc()
 		logutil.BgLogger().Info("the schema version is much older than the latest version", zap.Int64("currVer", currVer),
 			zap.Int64("latestSchemaVer", s.latestSchemaVer), zap.Reflect("deltas", newerDeltas))
-		return true
+		return nil, nil, true
 	}
+
+	changedTblMap := make(map[int64]uint64)
 	for _, item := range newerDeltas {
```

Comment:
Currently there are only 100 schema delta are saved, if a transaction last too long and more than 100 schema changes happen during that time, the transaction will fail.

Reply:
This is ensured by L166.

Reply:
if len(newerDeltas) == len(s.deltaSchemaInfos) {